proposals/0000-ghci-instances.rst

```diff
@@ -0,0 +1,146 @@
+
+Motivation
+------------
+
+Typeclass resolution and deriviation are integral parts of modern Haskell. Unfortunately, the compiler doesn't volunteer a lot of information on them during development. Figuring out whether a type has a specific instance usually requires you to work backwards: starting from the typeclass and going through every instance in Haddock, recursively, to figure out if your type satisfies it. This experience isn't optimal and can get annoying for large or nested types and classes.
+
+This is exactly the kind of task that the compiler could help us with though. Since it already has all the visible instances for a given piece of code loaded, we could just ask it to list all instances that a type matches! This proposal suggests adding a new command and related machinery to GHCi to acomplish this. An example of this command's possible execution is shown below:
+
+.. code-block:: none
+
+  >> import Data.Functor.Sum (Sum)
+  >>
+  >> :instances Sum [] []
+  Eq   a => Eq   (Sum [] [] a)
+  Ord  a => Ord  (Sum [] [] a)
+  Show a => Show (Sum [] [] a)
+  Read a => Read (Sum [] [] a)
+  Functor  (Sum [] [])
+  Foldable (Sum [] [])
+  Eq1      (Sum [] [])
+  Ord1     (Sum [] [])
+  Show1    (Sum [] [])
+  Read1    (Sum [] [])
+  FunctorWithIndex     (Either Int Int) (Sum [] [])
+  FoldableWithIndex    (Either Int Int) (Sum [] [])
+  TraversableWithIndex (Either Int Int) (Sum [] [])
+
+It would provide a new tool for Haskell developers, both new and experienced that helps during during development and debugging. It allows a new way of talking with our beloved compiler and to better understand how it thinks.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal adds a new command to GHCi called ``:instances`` which provides a listing of all valid instances for a given type.
+
+Syntax
+~~~~~~
+
+Valid queries to ``instances`` are types, if anything other than a type is provided as an argument, the command will return an error explaining the intended usage. The command interprets everything after ``instances`` as a single type, this means there is no way to specify multiple types to lookup instances of MPTCs. Holes can be used to represent free variables in the argument type, for example: ``Either _ _``. To express relations between multiple holes, named holes can be used, for example:: ``Either _a _a``.
+
+Execution
+~~~~~~~~~
+
+Provided with a valid type, ``instances`` will attempt to match it against the heads of all visible class instances and satisfy all the implied constraints. The output will consist of a formatted listing of all matching and satisfiable instances. Each instance should be simplified as much as possible, meaning that if an instance: ``(c ~ Bool) => C c`` were found it would be presented as ``C Bool``.
```

Comment:
I don't know if I worded this properly, but the intent is to say that given:
```
:instances T
```
We look for all unifiable instances `C T`,  for all visible classes `C`.

Reply:
Thanks for making the changes.  Just one more clarification:  the part about having to satisfy all implied constraints seems a little fishy to me.   If I write:

    :instances (Maybe _)

I would expect to see the `Eq` instance, correct?   The instance looks like this:

     instance Eq a => Eq (Maybe a)

When deciding if it should display it, GHC will unify the head against my query, which will succeed.  However, we can't simplify the context any further, as we are left with `Eq a` and we don't know how to simplify this.

OTOH, if I'd written `:instances (Maybe (_ -> _))`, then I probably wouldn't want to see that instance as there aren't any instances that match `Eq (_ -> _)`.

So we need some sort of a rule that will tell us which left-over constraints are OK, as in the first example, vs. not OK as in the second.

My suggestion would be that an instance should be displayed if we can unify with the head, and then simplify the instance context to constraints that only involve variables.  I haven't thought very deeply about that, so please take it with a grain of salt.

Reply:
> OTOH, if I'd written :instances (Maybe (_ -> _)), then I probably wouldn't want to see that instance as there aren't any instances that match Eq (_ -> _).

Wouldn't this just be a matter of going through the normal instance resolution process? When we attempt to reduce the context GHC should tell us that there is no instance for `Eq (_ -> _)`?

Reply:
Hmm, yea I see the problem now, I'll add that specific example and reword that sentence. What we'd like is effectively that only constraints on the type-holes remain after simplifying.